Add mouse-wheel support to Dolly and FlyThrough navigation modes.

biagas opened the pull request:
### Description
These resolve #2727:
* Added OnMouseWheelForward and OnMouseWheelBackward methods to Dolly and FlyThrough interaction modes.
* Added support methods to VisitInteractor:  `DollyCameraTowardFocus3D(double f)` (for Dolly)  and  `DollyCameraAndFocus3D(double)` (for FlyThrough)
* Fixed setting of view3D focus in 'Flythrough' mode, so that multiple calls to ResetView will no longer cause the plot to disappear.
* Automatically reset view on Navigation mode changes in 3D, since each mode handles the camera differently.

Resolves #17465:

* Fixed problem with right-clicking in Viewer window to access popup menu preventing mouse-wheel interaction.

Resolves #17510:

* Fix bug that made plots disappear from view when adding plots in Flythrough mode.

Resolves #17496.

* Ensure bounding box resets after 'Scale 3D axes' turned off.

Resolves #17469 and #17511:
* Make Recenter view be a no-op for Flythrough mode.



### Type of change

<!-- Please check one of the boxes below -->

* [X] Bug fix
* [ ] New feature
* [ ] Documentation update
* [ ] Other <!-- please explain with a note below -->

### How Has This Been Tested?

Used mouse wheel successfully to zoom in Dolly and FlyThrough navigation modes.

### Checklist:

<!-- For items in this checklist that do not apply, simply insert two tilde chars, `~~`, just ahead of the left bracket char, `[` at the beginning of a line. Each line ends with two tilde chars to make doing such ~~strikeouts~~ easy. -->

- [X] I have followed the [style guidelines][1] of this project.~~
- [X] I have performed a self-review of my own code.~~
- [X] I have commented my code where applicable.~~
- [X] I have updated the release notes.~~
~~- [ ] I have made corresponding changes to the documentation.~~
~~- [ ] I have added debugging support to my changes.~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works.~~
~~- [ ] I have confirmed new and existing unit tests pass locally with my changes.~~
~~- [ ] I have added new baselines for any new tests to the repo.~~
~~- [ ] I have NOT made any changes to [*protocol* or *public interfaces*][3] in an RC branch.~~
- [X] I have assigned reviewers (see [VisIt's PR procedures][2] for more information).~~

[1]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/StyleGuide.html
[2]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/pr_create.html#reviewers
[3]: https://visit-sphinx-github-user-manual.readthedocs.io/en/develop/dev_manual/RCDevelopment.html#communication-protocols-and-public-apis

biagas committed:
Add mouse-wheel support to Dolly and FlyThrough navigation modes.

biagas committed:
Update release notes.

biagas commented:
@markcmiller86 This may resolve parts of #2727.

biagas committed:
Merge branch 'develop' into bugfix/biagas/add_mousewheel_support_dolly_flythrough

markcmiller86 commented:
> @markcmiller86 This may resolve parts of #2727.

Was this an appeal for me to try to test this PR and see if it does?

biagas commented:
@markcmiller86 Yes, I would like to know this works on Mac, I believe it should, since mousewheel in Navigation mode works.
I ran into the problem myself (looking into navigation problems with VTK-9) so I looked to see if any tickets were already in the system.  While this PR doesn't address all the problems, it would be good to have it linked to show which portions have  been fixed.

biagas committed:
Merge branch 'develop' into bugfix/biagas/add_mousewheel_support_dolly_flythrough

biagas commented:
@markcmiller86 I found a fix for the flythrough rightclick->ResetView makes the plot disappear. I will add the fix to this PR and then maybe I can fully resolve #2727 :-)

biagas commented:
Spoke too soon :-( . I fixed a different problem with right-click reset view. Must have forgot to click 'apply' after changing to Fly Through mode in initial testing.

markcmiller86 commented:
Testing now. As an aside...am seeing something funky (could be my installation of third-party libs)...

```
% git rev-parse --abbrev-ref HEAD
bugfix/biagas/add_mousewheel_support_dolly_flythrough
% ./bin/visit
Running: gui -dv
Running: viewer -dv -geometry 1519x1174+401+3 -borders 20,0,0,1 -shift 0,0 -preshift 0,0 -defer -host 127.0.0.1 -port 5600
Running: mdserver -dv -host 127.0.0.1 -port 5601
dlerror() = dlopen(libospray_module_ispc.dylib, 10): image not found
dlerror() = dlopen(libospray_module_ispc_native_float.dylib, 10): image not found
dlerror() = dlopen(libospray_module_ispc_avx2_float.dylib, 10): image not found
dlerror() = dlopen(libospray_module_ispc_avx_float.dylib, 10): image not found
dlerror() = dlopen(libospray_module_ispc_sse4_float.dylib, 10): image not found
#osp: INITIALIZATION ERROR --> could not open module lib ospray_module_ispc
```

though VisIt appears to still start and work correctly.

markcmiller86 commented:
Ok, it works mostly I think :tada: :heart: Attached a short movie of behavior I am seeing...


https://user-images.githubusercontent.com/5720676/158887279-480123b4-8031-456a-8cd4-82168fca6f40.mov

biagas committed:
Fix more navigation issues.

Fixed problem where multiple 'ResetView's would make the plot disappear in 'Flythrough' navigation mode.
Fixed problem where right-clicking in the viewer windows to access the popup menu would prevent mouse-wheel navigation until another mouse-click was performed.

biagas commented:
Am I right in that the main issue displayed by your video was the plot disappearing when calling ResetView in Flythrough mode?
It should be resolved now with my latest commit to this PR.

markcmiller86 commented:
> Am I right in that the main issue displayed by your video was the plot disappearing when calling ResetView in Flythrough mode?

I don't think that is all the problems that are there. Do you get same behavior as I was getting?

- Each roll/click of the wheel as lesser and lesser effect in dolly mode as you dive **in** to the view. That doesn't seem right. Or, maybe I don't understand what *dollying* really means.
- Scale of roll/clicks of the wheel seems off in flythrough mode. I only get 5 steps and I've flown through the whole seen.
- Repeatedly hitting reset view button changes focus (y-value) continuously increasing it.

biagas commented:
> * Each roll/click of the wheel as lesser and lesser effect in dolly mode as you dive **in** to the view. That doesn't seem right. Or, maybe I don't understand what _dollying_ really means.

I noticed this too, but if I click 'Reset View' while in Dolly Mode the mouse wheel seems to behave better. Is that true for you, too?

> * Scale of roll/clicks of the wheel seems off in flythrough mode. I only get 5 steps and I've flown through the whole seen.

Do you get better behavior after a 'Reset View' ?

> * Repeatedly hitting reset view button changes focus (y-value) continuously increasing it.

This is for Flythrough mode, correct?  Before my latest commit?  This should be fixed with latest commit to the branch for this PR.

I'm beginning to think there needs to be a slight manipulation of the View when changing Navigation modes, because there is different behavior *after* a reset view!

biagas commented:
> I'm beginning to think there needs to be a slight manipulation of the View when changing Navigation modes, because there is different behavior _after_ a reset view!

Actually, I think coming in/out of Flythrough requires a ResetView so that the other modes work correctly.

markcmiller86 commented:
> > * Repeatedly hitting reset view button changes focus (y-value) continuously increasing it.
> 
> This is for Flythrough mode, correct? Before my latest commit? This should be fixed with latest commit to the branch for this PR.

I confirm your most recent change fixes that.

> I'm beginning to think there needs to be a slight manipulation of the View when changing Navigation modes, because there is different behavior _after_ a reset view!

I think I agree with that too. I tried something simple and cycled through each navigation mode and hit RestView between each. Between FlyThrough and either of the two others, the view changes a bit.

> * Scale of roll/clicks of the wheel seems off in flythrough mode. I only get 5 steps and I've flown through the whole seen.

I am unable to say for sure but after an appropriate sequence of view changes and reset view operations, FlyThrough begins to behave more smoothly and with much finer granularity of control that what the above bullet described too.

markcmiller86 commented:
I happend to be rendering Mesh plots and I see wildly varying quality of mesh plots as I vary view parameters in FlyThrough or Dolly mode. See below some various images...

![Screen Shot 2022-03-17 at 7 02 01 PM](https://user-images.githubusercontent.com/5720676/158923693-29179679-21f1-4f14-81c7-05babd3e0f4b.png)
![Screen Shot 2022-03-17 at 7 01 52 PM](https://user-images.githubusercontent.com/5720676/158923695-fa619dbb-3b41-4e58-86c5-144923cb25b9.png)
![Screen Shot 2022-03-17 at 7 03 36 PM](https://user-images.githubusercontent.com/5720676/158923689-465b36e2-5dd6-4a33-8be8-ed74c280e3f2.png)
![Screen Shot 2022-03-17 at 7 02 06 PM](https://user-images.githubusercontent.com/5720676/158923691-f314e014-1976-4b82-8ee5-e9f1c257b9ec.png)

biagas committed:
Automatically reset view on Interactor navigation mode changes in 3D.

Also changed MotionFactor adjustment for FlyThrough's mouse wheel interaction so it more closely matches Middle-mouse.

biagas committed:
Merge branch 'develop' into bugfix/biagas/add_mousewheel_support_dolly_flythrough

biagas committed:
Fix autmatic view reset so ViewAttributes are updated as well.

Update release notes.

biagas committed:
Merge branch 'bugfix/biagas/add_mousewheel_support_dolly_flythrough' of https://github.com/visit-dav/visit into bugfix/biagas/add_mousewheel_support_dolly_flythrough

biagas committed:
Merge branch 'develop' into bugfix/biagas/add_mousewheel_support_dolly_flythrough

biagas committed:
Fix AdjustView3d for Flythrough mode.

Solves view strangeness in Flythrough mode when new plots with are added.

biagas committed:
Merge branch 'develop' into bugfix/biagas/add_mousewheel_support_dolly_flythrough

biagas committed:
Ensure bounding box is reset when 'Scale 3D axes' is turned off.

Resolves #17496

biagas committed:
Add more of the navigation-interaction fixes to the release notes.

biagas committed:
Merge branch 'develop' into bugfix/biagas/add_mousewheel_support_dolly_flythrough

biagas committed:
Update docs on Interactors to  indicate automatic reset of View.

biagas committed:
Merge branch 'bugfix/biagas/add_mousewheel_support_dolly_flythrough' of https://github.com/visit-dav/visit into bugfix/biagas/add_mousewheel_support_dolly_flythrough

biagas committed:
Make RecenterView be a no-op in Flythrough mode.

biagas committed:
Merge branch 'develop' into bugfix/biagas/add_mousewheel_support_dolly_flythrough

biagas committed:
Merge branch 'develop' into bugfix/biagas/add_mousewheel_support_dolly_flythrough

biagas committed:
Merge branch 'develop' into bugfix/biagas/add_mousewheel_support_dolly_flythrough

biagas committed:
Merge branch 'develop' into bugfix/biagas/add_mousewheel_support_dolly_flythrough

biagas commented:
@markcmiller86 do you have time to review this?

markcmiller86 commented:
Sure thing. I justed updated and am rebuilding locally.

markcmiller86 commented:
Should `ResetView` mean the same thing in all three modes? It doesn't. At least not with `multi_ucd3d.silo`. If I switch between the modes and hit `ResetView` I get changing views between `Trackball` and `Flythrough`.

I tried to use `Trackball` to orient `multi_ucd3d.silo` so that its being viewed from the send (instead of the end) and then wanted to fly through it in `Flythrough` mode. That seems like a reasonable thing to do. When I set to `Flythrough`, the view I set up in `Trackball` mode got lost and it got reset to end-on view (though a different zoom than the end-on view I would get after resetting view in `Trackball` mode).

markcmiller86 commented:
> I tried to use `Trackball` to orient `multi_ucd3d.silo` so that its being viewed from the send

send ---> side

markcmiller86 commented:
BTW...I forget what we decided regarding view continuity between the modes but maybe we decided its fine if `ResetView` winds up causing discontinuities as interaction mode is varied.

markcmiller86 commented:
Mesh lines appear to be doing a lot better than before 🎉

biagas commented:
When changing modes, there is an automatic 'ResetView' (new behavior) is that what you are talking about?  I made this change because each mode treats the views fundamentally differently.

markcmiller86 commented on src/doc/using_visit/VisualizationWindows/Interactor_settings.rst:
```diff
@@ -21,47 +18,27 @@ that have interactor properties. The **Interactors window** is shown in
-the edges of the plots' bounding box when the vis window is in 2D mode. If
-you want to turn off the guidelines, click off the **Show Guidelines** check
-box in the **Interactors window**.
+The zoom interactor settings are mostly used when the vis window is in zoom mode. 
```
Using `vis window` (here and a few other places). Is that really user-facing terminology? I don't think so. Maybe best to use `viewer` or just `window` (e.g. drop the `vis` part).

biagas replied:
I didn't write that section. I just changed to single line per sentence.

biagas replied:
Is there a way to capture your suggestions to a new ticket?

markcmiller86 replied:
> I didn't write that section. I just changed to single line per sentence.

I suppose I should have been able to pick up on that in examining the diffs. But, it hadn't occurred to me and so I am finding myself spending time looking at material you don't want reviewed.

Can you suggest anything so I can avoid this in future?

biagas replied:
I try to add notes to the effect that I changed to single-line per sentence. It may have been part of one of my commit messages, or I may be thinking of a different PR. At any rate, if it had been a note in my PR I think it could have been helpful. My apologies for this oversight

markcmiller86 replied:
> Is there a way to capture your suggestions to a new ticket?

There is something that comes close. In the elipsis in upper-right of the comment, pull down to `Reference in new issue`. That brings up a new dialog (which is pretty limited in functionality) but does allow you to create a new ticket based on the original text. I just did that for the case below :wink:

markcmiller86 replied:
> My apologies for this oversight

Not necessary. I was just looking for a suggested tweek to our practices.

Let me propose this...when you wanna make a pass fixing line length *globally* in a `.rst` file (except `functions.rst` :wink:) as opposed to just the paragraph you are there to edit, do that and *only* that work in a separate commit. Maybe put that commit *after* whatever *real* changes you are there to make. The first diff is the real work. The second diff can be ignored.

markcmiller86 commented on src/doc/using_visit/VisualizationWindows/Interactor_settings.rst:
```diff
@@ -21,47 +18,27 @@ that have interactor properties. The **Interactors window** is shown in
-VisIt to always force a square zoom rectangle so that you don't have to
-use the *Shift* key, you can click on the **Clamp to Square** check box,
-click **Apply** in the **Interactors window** and save your settings.
+When sweeping out a zoom rectangle in zoom mode, VisIt allows you to draw a rectangle of any proportion. 
```
Explaining this behavior of the aspect ratio of the zoom box is great to include in our docs.

I am not seeing this behavior on macOS though. Here is what I do see (with a 2D dataset, `curve2d.silo`)...

- If I do depress `Shift` before dragging while in `Zoom` mode, I get panning...which is same as in `Navigate` mode.
- If I depress `Shift` *after* beginning a zoom drag, I get no effect (no constraining)
- I tried other keys as well (`Ctrl`, `Alt`, and `Command` (or the Apple option key). For `Command` I get an inverted zoom event (e.g. it zooms out not in ;)
- I though I recall seeing a constrained drag zoom on macOS in the past but am now unable to do it.

markcmiller86 commented on src/doc/using_visit/VisualizationWindows/Interactor_settings.rst:
```diff
@@ -21,47 +18,27 @@ that have interactor properties. The **Interactors window** is shown in
-use the *Shift* key, you can click on the **Clamp to Square** check box,
-click **Apply** in the **Interactors window** and save your settings.
+When sweeping out a zoom rectangle in zoom mode, VisIt allows you to draw a rectangle of any proportion. 
+The relative shape of the zoom rectangle, in turn, influences the shape of the viewport drawn in the vis window. 
```
I think `aspect ratio` is maybe a better term than `relative shape`. I think we use `aspect ratio` elsewhere both in the GUI and in the docs to mean same thing.

markcmiller86 commented on src/doc/using_visit/VisualizationWindows/Interactor_settings.rst:
```diff
@@ -21,47 +18,27 @@ that have interactor properties. The **Interactors window** is shown in
-order to make better use of the vis window's pixels.
+Using the **Clamp to Square** zoom mode is a good way to maximize the amount of the vis window that is used when you zoom in on plots and when the vis window is in zoom mode. 
+When the vis window is in navigate mode, the middle mouse button also effects a zoom. 
+By default, zooming with the middle mouse button zooms into the plots but keeps the same vis window viewport which may, depending on the aspect ratio of the plots, not make the best use of the vis window's pixels. 
```
```suggestion
By default, dragging up and down with the middle mouse button zooms in and out, respectively, but keeps the same aspect ratio which may, depending on the situation, not make the best use of the vis window's pixels. 
```

markcmiller86 commented on src/doc/using_visit/VisualizationWindows/Interactor_settings.rst:
```diff
@@ -21,47 +18,27 @@ that have interactor properties. The **Interactors window** is shown in
-navigation style moves the camera and allows you to fly into plots and
-out the other side.
+When VisIt displays 3D plots, there are a few navigation styles from which you can choose by clicking on the following radio buttons in the **Interactors window**: **Trackball**, **Dolly**, and **Flythrough**.  
+The default navigation style for 3D plots is: Trackball and it allows you to interactively rotate plots and move around them but it keeps the camera at a fixed distance from the plots and while it can get infinitely close to plots when you zoom in, it can never touch them or go inside of them. 
```
```suggestion
The default navigation style for 3D plots is **Trackball**. This allows users to interactively rotate plots and move around them but it keeps the camera at a fixed distance from the plots. While it can get infinitely close to plots when zoomed in, it can never touch them or go inside of them. 
```

markcmiller86 commented on src/doc/using_visit/VisualizationWindows/Interactor_settings.rst:
```diff
@@ -21,47 +18,27 @@ that have interactor properties. The **Interactors window** is shown in
-out the other side.
+When VisIt displays 3D plots, there are a few navigation styles from which you can choose by clicking on the following radio buttons in the **Interactors window**: **Trackball**, **Dolly**, and **Flythrough**.  
+The default navigation style for 3D plots is: Trackball and it allows you to interactively rotate plots and move around them but it keeps the camera at a fixed distance from the plots and while it can get infinitely close to plots when you zoom in, it can never touch them or go inside of them. 
+The Dolly navigation style behaves like the trackball style except that the when the camera zooms, it is actually moved. 
```
```suggestion
The **Dolly** navigation style behaves like the trackball style except that *zooming* is affected by moving the camera closer to the plots, not by changing the angle of view. 
```

markcmiller86 commented on src/doc/using_visit/VisualizationWindows/Interactor_settings.rst:
```diff
@@ -21,47 +18,27 @@ that have interactor properties. The **Interactors window** is shown in
+When VisIt displays 3D plots, there are a few navigation styles from which you can choose by clicking on the following radio buttons in the **Interactors window**: **Trackball**, **Dolly**, and **Flythrough**.  
+The default navigation style for 3D plots is: Trackball and it allows you to interactively rotate plots and move around them but it keeps the camera at a fixed distance from the plots and while it can get infinitely close to plots when you zoom in, it can never touch them or go inside of them. 
+The Dolly navigation style behaves like the trackball style except that the when the camera zooms, it is actually moved. 
+The Flythrough navigation style moves the camera and allows you to fly into plots and out the other side.
```
```suggestion
The **Flythrough** navigation style moves the camera and allows you to fly into plots and out the other side.
```

markcmiller86 reviewed (approved):
This is great. I think its in way better shape than it was before.

I did have some observations and some suggestions. Take 'em or leave 'em.
